web/src/user_data.ts

```diff
@@ -0,0 +1,98 @@
+import * as channel from "./channel.ts";
+import type {StateData} from "./state_data";
+import {user_schema} from "./state_data.ts";
+
+export const response_schema = z.object({
+    members: z.array(user_schema),
+    result: z.string(),
+    msg: z.string(),
+});
+
+type UsersFetchResponse = z.infer<typeof response_schema>;
+
+let valid_user_ids = new Set<number>();
+type FetchUserDataParams = {
+    user_ids: string;
+    client_gravatar?: boolean;
+    include_custom_profile_fields?: boolean;
+    success?: (users: UsersFetchResponse["members"]) => void;
+    error?: (xhr?: JQuery.jqXHR) => void;
+};
+
+export function populate_valid_user_ids(params: StateData["user_groups"]): void {
+    for (const user_group of params.realm_user_groups) {
+        if (user_group.is_system_group) {
+            valid_user_ids = valid_user_ids.union(new Set(user_group.members));
+        }
+    }
+}
+
+export function get_valid_user_ids(): Set<number> {
+    return valid_user_ids;
+}
+
+export function is_valid_user_id(user_id: number): boolean {
+    return valid_user_ids.has(user_id);
+}
+
+export function fetch_users_from_server(opts: FetchUserDataParams): void {
+    const params = {
+        user_ids: opts.user_ids,
+        client_gravatar: opts.client_gravatar,
+        include_custom_profile_fields: opts.include_custom_profile_fields,
+    };
+
+    channel.get({
+        url: "/json/users",
+        data: params,
+        success(data) {
+            if (opts.success) {
+                const members = response_schema.parse(data).members;
+                opts.success(members);
+            }
+        },
+        error(xhr: JQuery.jqXHR<unknown>) {
+            if (opts.error) {
+                opts.error(xhr);
+            }
+        },
+    });
+}
+
+export async function get(user_ids: number[]): Promise<UsersFetchResponse["members"]> {
+    assert(user_ids.length > 0, "user_ids should not be empty.");
+    const invalid_user_ids = new Set(user_ids).difference(get_valid_user_ids());
+    if (invalid_user_ids.size > 0) {
+        blueslip.info("Ignored invalid user_ids" + [...invalid_user_ids].join(", "));
+    }
+    const valid_user_ids = get_valid_user_ids().intersection(new Set(user_ids));
+    if (valid_user_ids.size === 0) {
+        return [];
+    }
+    return new Promise((resolve, reject) => {
+        fetch_users_from_server({
+            user_ids: JSON.stringify([...valid_user_ids]),
+            success(users) {
+                resolve(users);
+            },
+            error(xhr) {
+                let error_message = "Failed to fetch users.";
+                if (xhr) {
+                    const error = z.object({msg: z.string().optional()}).parse(xhr.responseJSON);
+                    if (error.msg) {
+                        error_message = error.msg;
+                    }
+                }
+                reject(new Error(error_message));
+            },
+        });
+    });
+}
```

Comment:
I'm not sure this separate module makes sense as an abstraction; it might be better for this to just be part of `people.ts`, since a lot of `people.ts` helpers may want to interact with the `valid_user_ids` structure. But if you like it, I think `user_fetch.ts` might be a more appropriate name.

Reply:
Yeah, makes sense, moved the code to `people.ts`.